core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/FileAsyncResponseTransformer.java

```diff
@@ -44,13 +49,48 @@
     private volatile AsynchronousFileChannel fileChannel;
     private volatile CompletableFuture<Void> cf;
     private volatile ResponseT response;
+    private final long position;
+    private final FileWriteOption fileModificationOption;
+    private final boolean deleteOnFailure;
 
     public FileAsyncResponseTransformer(Path path) {
         this.path = path;
+        this.fileModificationOption = FileWriteOption.CREATE_NEW;
+        this.deleteOnFailure = true;
+        this.position = 0L;
+    }
+
+    public FileAsyncResponseTransformer(Path path, FileTransformerConfiguration fileConfiguration) {
+        this.path = path;
+        this.fileModificationOption = fileConfiguration.fileModificationOption().orElse(FileWriteOption.CREATE_NEW);
+        this.position = determineFilePositionToWrite(path);
+        this.deleteOnFailure = fileConfiguration.deleteOnFailure().orElse(true);
+    }
+
+    private long determineFilePositionToWrite(Path path) {
+        if (fileModificationOption == CREATE_OR_APPEND_EXISTING) {
+            try {
+                return Files.size(path);
+            } catch (NoSuchFileException e) {
+                // Ignore
+            } catch (IOException exception) {
+                throw SdkClientException.create("Cannot determine the current file size " + path, exception);
+            }
```

Comment:
This seems like an expensive thing to do all the time (e.g. if the file is always being created for the first time). Is there any way we can defer this until we need it or use a solution that doesn't rely on catching `NoSuchFileException`?

Reply:
Hmm couldn't we do it at the time we open the file?

Reply:
No not really, we open the file in `onStream` method, so the position we get at this point may not be correct if the previous attempt had written a few bytes already before it failed.